server/src/main/java/org/elasticsearch/action/search/SearchContextId.java

```diff
@@ -0,0 +1,122 @@
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.search.SearchShardTarget;
+import org.elasticsearch.search.internal.AliasFilter;
+import org.elasticsearch.search.internal.ShardSearchContextId;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class SearchContextId {
+    private final Map<ShardId, SearchContextIdForNode> shards;
+    private final Map<String, AliasFilter> aliasFilter;
+
+    private SearchContextId(Map<ShardId, SearchContextIdForNode> shards, Map<String, AliasFilter> aliasFilter) {
+        this.shards = shards;
+        this.aliasFilter = aliasFilter;
+    }
+
+    public Map<ShardId, SearchContextIdForNode> shards() {
+        return shards;
+    }
+
+    public Map<String, AliasFilter> aliasFilter() {
+        return aliasFilter;
+    }
+
+    static String encode(List<SearchPhaseResult> searchPhaseResults, Map<String, AliasFilter> aliasFilter, Version version) {
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            out.writeVInt(version.id);
+            out.writeVInt(searchPhaseResults.size());
+            for (SearchPhaseResult searchPhaseResult : searchPhaseResults) {
+                SearchShardTarget target = searchPhaseResult.getSearchShardTarget();
+                ShardId shardId = target.getShardId();
+                // shardId
+                out.writeString(shardId.getIndex().getName());
+                out.writeString(shardId.getIndex().getUUID());
+                out.writeVInt(shardId.getId());
+                // nodeId
+                if (target.getClusterAlias() != null) {
+                    out.writeByte((byte) 1);
+                    out.writeString(target.getClusterAlias());
+                } else {
+                    out.writeByte((byte) 0);
+                }
+                out.writeString(target.getNodeId());
+                // readerId
+                out.writeString(searchPhaseResult.getContextId().getReaderId());
+                out.writeLong(searchPhaseResult.getContextId().getId());
+            }
+            out.writeMap(aliasFilter, StreamOutput::writeString, (outstream, val) -> val.writeTo(outstream));
+            return Base64.getUrlEncoder().encodeToString(BytesReference.toBytes(out.bytes()));
+        } catch (IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    static SearchContextId decode(String id) {
+        final ByteBuffer byteBuffer;
+        try {
+            byteBuffer = ByteBuffer.wrap(Base64.getUrlDecoder().decode(id));
+        } catch (Exception e) {
+            throw new IllegalArgumentException("invalid id: [" + id + "]", e);
+        }
+        try (StreamInput in = new ByteBufferStreamInput(byteBuffer)) {
+            Version version = Version.fromId(in.readVInt());
```

Comment:
Do we intend to use this `version` parameter as well?

Reply:
Yes. With the version, we can add new features to the search context id without breaking BWC.